lightning/src/routing/network_graph.rs

```diff
@@ -668,6 +685,105 @@ impl_writeable_tlv_based!(ChannelInfo, {
 	(12, announcement_message, required),
 });
 
+/// A wrapper around [`ChannelInfo`] representing information about the channel as directed from a
+/// source node to a target node.
+pub struct DirectedChannelInfo<'a: 'b, 'b> {
+	channel: &'a ChannelInfo,
+	direction: Option<&'b DirectionalChannelInfo>,
+	source: &'b NodeId,
+	target: &'b NodeId,
+}
+
+impl<'a: 'b, 'b> DirectedChannelInfo<'a, 'b> {
+	/// Returns the node id for the source.
+	pub fn source(&self) -> &'b NodeId { self.source }
+
+	/// Returns the node id for the target.
+	pub fn target(&self) -> &'b NodeId { self.target }
+
+	/// Consumes the [`DirectedChannelInfo`], returning the wrapped parts.
+	pub fn into_parts(self) -> (&'a ChannelInfo, Option<&'b DirectionalChannelInfo>) {
+		(self.channel, self.direction)
+	}
+
+	/// Returns the [`EffectiveCapacity`] of the channel in a specific direction.
+	///
+	/// This is either the total capacity from the funding transaction, if known, or the
+	/// `htlc_maximum_msat` for the direction as advertised by the gossip network, if known,
+	/// whichever is smaller.
+	pub fn effective_capacity(&self) -> EffectiveCapacity {
+		Self::effective_capacity_from_parts(self.channel, self.direction)
+	}
+
+	/// Returns the [`EffectiveCapacity`] of the channel in the given direction.
+	///
+	/// See [`Self::effective_capacity`] for details.
+	pub fn effective_capacity_from_parts(
+		channel: &ChannelInfo, direction: Option<&DirectionalChannelInfo>
+	) -> EffectiveCapacity {
+		let capacity_msat = channel.capacity_sats.map(|capacity_sats| capacity_sats * 1000);
+		direction
+			.and_then(|direction| direction.htlc_maximum_msat)
+			.map(|max_htlc_msat| {
+				let capacity_msat = capacity_msat.unwrap_or(u64::max_value());
+				if max_htlc_msat < capacity_msat {
+					EffectiveCapacity::MaximumHTLC { amount_msat: max_htlc_msat }
+				} else {
+					EffectiveCapacity::Total { capacity_msat }
```

Comment:
I don't think this is possible because `max_htlc_msat` must be smaller than that. BOLT 7 requires that it is less than or equal to the channel capacity, and we make sure it is less than `MAX_VALUE_MSAT`.